lib/winapi/profiling.c

```diff
@@ -10,28 +10,19 @@
 #include <stddef.h>
 #include <xboxkrnl/xboxkrnl.h>
 
-BOOL QueryPerformanceCounter (LARGE_INTEGER *lpPerformanceCount)
-{
-    assert(lpPerformanceCount != NULL);
-
-    lpPerformanceCount->QuadPart = __rdtsc();
-    return TRUE;
-}
-
-BOOL QueryPerformanceFrequency (LARGE_INTEGER *lpFrequency)
-{
-    assert(lpFrequency != NULL);
-
 #ifdef USE_RDTSC_FOR_FREQ
-    #define AVG_SET 10
-    ULARGE_INTEGER f_rdtsc, avg;
-    ULONG f_ticks = 0;
+static LARGE_INTEGER frequency = {0, 0};
+static void __attribute__((constructor)) PrimeQueryPerformanceFrequency ()
+{
+    #define AVG_SET 2
+    ULARGE_INTEGER f_rdtsc, avg = {0, 0}, s_rdtsc;
+    ULONG f_ticks = 0, s_ticks = 0;
 
-    avg.QuadPart = 0;
+    Sleep(500);
 
     for (int i = 0; i < AVG_SET; i++) {
-        ULARGE_INTEGER s_rdtsc;
-        ULONG s_ticks;
+        // If we call rdtsc too fast we'll end up with div by 0
+        Sleep(200);
```

Comment:
ofc we should prefer not having to wait at all, which should be possible when reading the CPU clockspeed from wherever it is stored.

Reply:
> ofc we should prefer not having to wait at all, which should be possible when reading the CPU clockspeed from wherever it is stored.

We can't guarantee this is accurate. The old method of changing the FSB on CPU replaced Xbox' were to mess with the clock generator output, not the hardware register used for that.

Reply:
I'd argue "CPU replaced Xbox" is no longer an Xbox and out of scope for nxdk.
In theory, one could also make a "GPU replaced Xbox" etc. - one could change any clock on the system.

It's impossible for us to support all of these variations in a useful way.

nxdk is open-source, so if people want to support non-stock clocks, then they can keep around a patched nxdk to make special builts.
I guess we could mark timing related variables (in code or in binary for hex editing.. at potential performance cost) in the nxdk libs, but even then any application might make its own assumptions when running on Xbox, so you'd probably still have to bring your own hex edits or patches for closed-source apps.

There's no good solution for supporting Xbox-like platforms in upstream nxdk.

Reply:
> It's impossible for us to support all of these variations in a useful way.

Whether or not it's possible to support all variations is moot. I'm attempting to support **this** specific thing, and I hate how much I have to try to justify my changes to you and others. Is there such a risk providing an *accurate to now* clock rate instead of a hard coded value?

> In theory, one could also make a "GPU replaced Xbox" etc.

These have been done while you've been gone. Replacing the GPU from 1.6 to say 1.0, etc.

Reply:
> Replacing the GPU from 1.6 to say 1.0, etc.

It's still the same GPU. I'm talking about plugging some Voodoo 3dfx chip inside an Xbox.

> Is there such a risk providing an accurate to now clock rate instead of a hard coded value?

No. But:
- There's a huge risk in startup delay. (application breaking! fixable)
- The result you get is, despite the huge delay, more than 100x worse. (application breaking! fixable)
- This not being a constant makes it impossible to do constant propagation at build-time. (potentially application breaking. not fixable)

... and all of this for a niche use case that's not obtainable with a "stock" Xbox.. regardless of how badly it's damaged.

All of this matters *a lot*.

Reply:
> Slightly increased application size

100 bytes? Are we on an 8086? If this were a org standard we'd have inline assembly all over the place.

> The result you get is, despite the huge delay, more than 100x worse.

To quote you...
```
That's a 0.00013636369% error.. how is that not negligible?
```

> This not being a constant makes it impossible to do constant propagation at build-time. (potentially application breaking. not fixable)

Application *shouldn't* assume about the hardware. We're providing a tool (local box variable timing) to avoid, *bad assumptions*.

> There's a huge risk in startup delay. (application breaking! fixable)

I don't disagree. How long are you willing to Sleep on startup?